feat(new-post): Add 'is tdd a gimmick?'

jLouzado opened the pull request.

jLouzado committed:
feat(new-post): Add 'is tdd a gimmick?'

dlebrero commented on content/blog/2019-10-3-tdd-gimmick/index.md:
```diff
@@ -0,0 +1,196 @@
+
+# Statement
+
+Every day we see new frameworks rise and fall, and a certain amount of skepticism is healthy to keep from jumping on every new thing that comes along.
```
I wouldn't call TDD new. It has been around for at least 18 years in its current form.

dlebrero commented on content/blog/2019-10-3-tdd-gimmick/index.md:
```diff
@@ -0,0 +1,196 @@
+
+If we have coverage we can have confidence, and with that confidence you get some effects:
+
+- code comprehension will always take linear time even if code-complexity increases exponentially.
```
I personally think this is a bold statement. You still need to understand the "production" code, that if it is complex, it will take a long time to understand. I wouldn't talk about "linear".

What it is very true is that tests will let you understand it better, both because of the documentation that they provide and because, as you mention above, you can "play" with the code (by changing/deleting lines) and see how it affects the test suite.

dlebrero commented on content/blog/2019-10-3-tdd-gimmick/index.md:
```diff
@@ -0,0 +1,196 @@
+If we have coverage we can have confidence, and with that confidence you get some effects:
+
+- code comprehension will always take linear time even if code-complexity increases exponentially.
+  - the tests will always exhaustively describe the function's behaviour, regardless of how "clever" the implementation is.
```
Nowadays, when I see "exhaustively" I think about property based testing, not in example based testing.

dlebrero commented on content/blog/2019-10-3-tdd-gimmick/index.md:
```diff
@@ -0,0 +1,196 @@
+  - compilation errors count as failures, and
+- Only write as much code as is required to fix that failure, nothing more.
+  - this is the `GREEN` phase
+- Rinse and repeat till all your requirements are implemented.
```
I do refactor as I implement, not at the end of writing all the tests. I do not refactor after each test if I do not consider it necessary, so I would personally move this under the refactor phase and make it obvious that the refactoring is optional.

dlebrero commented on content/blog/2019-10-3-tdd-gimmick/index.md:
```diff
@@ -0,0 +1,196 @@
+
+First, a quick recap of TDD's rules.
+
+- If you want to write new code (or make changes to an existing unit) first write a failing test.
```
I like to make it clear that you are supposed to write one and just one failing test. So I would highlight the "a" or change it for one and only one.

dlebrero commented on content/blog/2019-10-3-tdd-gimmick/index.md:
```diff
@@ -0,0 +1,196 @@
+- Rinse and repeat till all your requirements are implemented.
+- Go back over your implementation and clean-up.
+  - This is the `REFACTOR` phase.
+
```
Maybe make it clear on which steps do you run the test suite?

jLouzado replied:
is there a prescribed time to run tests? I just set them to run on every save.

dlebrero commented on content/blog/2019-10-3-tdd-gimmick/index.md:
```diff
@@ -0,0 +1,196 @@
+
+Writing code this way is definitely a little unnatural at first, but it makes sure that nothing extraneous ever enters your code base. When people talk about Lean-Testing, and only writing the minimum tests required, this is really what that looks like; no more tests than necessary, no more code than needed.
+
+Over time the intuition for this style will build and your compiler will be the best pair-programming partner you could ever ask for.
```
I think you are not serious about replacing pair programming with TDD, right? 🙈

dlebrero commented on content/blog/2019-10-3-tdd-gimmick/index.md:
```diff
@@ -0,0 +1,196 @@
+  - since your `expected` value will clearly be something different than the requirement, someone can flag it as an issue
+- once the error is spotted, it becomes easier to rewrite the bad test or to add tests for missing edge cases.
+
+Since refactoring is easier, this whole process happens much more quickly. And yes, while fixing one issue you're less likely to introduce a a regression even in this new feature.
```
Typo: "introduce a a regression"

dlebrero commented on content/blog/2019-10-3-tdd-gimmick/index.md:
```diff
@@ -0,0 +1,196 @@
+
+## This is all horribly complicated! I'm a pretty good programmer, I don't need all this
+
+I felt the same way when first starting out. Even now, when discussing some of the finer points of TDD it can feel a little overly-cautious and pessimistic. The thing is though, if there's more than 2 programmers working on a project it's just not possible to always be available for everyone.
```
It took me a while to digest why the second phrase ("The thing is ...") is here. After re-reading, I see that it has to do with the second phrase in the next paragraph ("In the absence..."). May I suggest to get both statements closer/together?

dlebrero commented on content/blog/2019-10-3-tdd-gimmick/index.md:
```diff
@@ -0,0 +1,196 @@
+
+Imagine you've written a function, and someone wants to make a change to it. If they knew that the tests were high-quality, they could just figure it out themselves and be fairly confident that they weren't doing anything wrong. In the absence of a test-suite they would (hopefully) come and sit on your head and work things out with you.
+
+In the worst case, they're under a lot of pressure so they make the change that re-introduces some subtle bug that you'd already taken care of initially. The bug might only get discovered months later, at which point maybe the dev who introduced the problematic-change is on leave and the commit message just says "fixed issue" so now you're really screwed.
```
It took me a while to understand this paragraph. I had to read the next to fully understand what you were referring to.

dlebrero commented on content/blog/2019-10-3-tdd-gimmick/index.md:
```diff
@@ -0,0 +1,196 @@
+
+In the worst case, they're under a lot of pressure so they make the change that re-introduces some subtle bug that you'd already taken care of initially. The bug might only get discovered months later, at which point maybe the dev who introduced the problematic-change is on leave and the commit message just says "fixed issue" so now you're really screwed.
+
+At it's best, TDD is a way to communicate your decisions without needing to write extensive documentation that no one will ever read (or maintain when they make changes). Good-quality unit tests save such an incredible amount of time, and ultimately that's more time that you have to work in peace.
```
I agree with this halfheartedly. Even if tests document the what, they do not document the why, which some times is what you need to decide if you need to fix the expectation of a test'

jLouzado replied:
Usually isn't the why just "because that was the requirement" ? So if the test description is written in terms of the product requirement then it should be easy to see what needs to be changed imo.

jLouzado replied:
Ref: https://github.com/jLouzado/patchcording.com/pull/11#issuecomment-541662742

dlebrero commented on content/blog/2019-10-3-tdd-gimmick/index.md:
```diff
@@ -0,0 +1,196 @@
+
+## Any others?
+
+Yes, if you're writing a new package, and there's two potential APIs that would both work equally fine but they both have different (say) performance implications based on their implementation then that might be a bad use-case for TDD.
```
Shouldn't be this a perfect case for TDD? The tests should be the same, the implementation can vary.

jLouzado replied:
made a change, let me know if this is clearer.

dlebrero commented on content/blog/2019-10-3-tdd-gimmick/index.md:
```diff
@@ -0,0 +1,196 @@
+
+## Are there any cases where one shouldn't use TDD?
+
+For application development, almost everything is a good candidate for TDD. The one exception for frontend development would be if your function returns a particular DOM structure, there's no point writing a test like `Assert <div>Hello World!</div>` and then having a function that just returns `<div>Hello World!</div>`.
```
I would put in this in the realm of tests that suffer from over specification, which comes to the question about what a good test is.

dlebrero commented on content/blog/2019-10-3-tdd-gimmick/index.md:
```diff
@@ -0,0 +1,196 @@
+
+In such a case you're really out near the edge of software-development and you're almost in research mode. Once you've sorted out architecture and explored the space sufficiently, that's when TDD might be of more use.
+
+I can't stress this enough though, almost everything related to Application Development can be done via TDD. As much as possible, extract your business logic into pure functions and TDD the heck out of it. You'll thank yourself later.
```
This is very good advice

jLouzado replied:
🙏

dlebrero commented on content/blog/2019-10-3-tdd-gimmick/index.md:
```diff
@@ -0,0 +1,196 @@
+
+At it's best, TDD is a way to communicate your decisions without needing to write extensive documentation that no one will ever read (or maintain when they make changes). Good-quality unit tests save such an incredible amount of time, and ultimately that's more time that you have to work in peace.
+
+## Are there any cases where one shouldn't use TDD?
```
As much as I love TDD, there are things that is not worth testing. I think you are right on saying "almost everything", which is also my point of view.

I wrote something regarding being too dogmatic at https://dev.to/danlebrero/the-tragedy-of-100-code-coverage

jLouzado replied:
great article... included the link as reference. :)

dlebrero commented on content/blog/2019-10-3-tdd-gimmick/index.md:
```diff
@@ -0,0 +1,196 @@
+
+## Cool, but I could just write the function and then write the test afterwards
+
+As far as I've spoken to people, no one disagrees with the need to have some unit tests. That means we all agree that it's possible to write buggy code. That implies that our unit-tests could _also_ have bugs; how would we detect that if we wrote it second?
```
I usually get the complain "if I write the test first, instead of after, I will need to keep updating the tests until I finish the implementation".

jLouzado replied:
this is a good one, I've tried to add an answer.

dlebrero reviewed (commented):
Thanks a lot for letting me review it. 

I always try to be extra careful with feedback, but if any comment somehow offend you, please do not take it seriously or personally!

Good luck with your TDD preaching! It is a hard one!

jLouzado committed:
fix(tdd): Remove implication that TDD is "new"

fix: pull/11#discussion_r332616917

jLouzado committed:
fix(tdd): clarify that refactoring is optional

fix: pull/11#discussion_r332621268

jLouzado committed:
feat(tdd): Highlight that only one failing test is allowed

fix: pull/11#discussion_r332622698

jLouzado committed:
fix(tdd): Add note to run tests on save

fix: pull/11#discussion_r332623874

jLouzado committed:
fix(tdd): Remove implication that TDD replaces pairing

fix: pull/11#discussion_r332625852

jLouzado committed:
fix(tdd): Remove typo

fix: pull/11#discussion_r332626855

jLouzado committed:
fix(tdd): Add "what are good tests" section

fix: pull/11#discussion_r332630772

jLouzado committed:
feat(tdd): Breakdown long para about TDD and scaling teams

fix: pull/11#discussion_r332628612, #discussion_r332629703, #discussion_r332630772

jLouzado committed:
feat(tdd): Added clarification for TDD and packages

fix: pull/11#discussion_r332632653

jLouzado committed:
feat(tdd): Include ref to when TDD is overkill

fix: pull/11#discussion_r332635097

jLouzado committed:
feat(tdd): Add specific example of immutability

fix: pull/11#discussion_r332638490

jLouzado committed:
refactor(tdd): Move the intro to tdd to the bottom

jLouzado committed:
fix(tdd): Remove claim about code-comprehension in linear time

jLouzado committed:
feat(tdd): language touch-ups

dlebrero commented on content/blog/2019-10-3-tdd-gimmick/index.md:
```diff
@@ -0,0 +1,283 @@
+If we have coverage we can have confidence, and with that confidence you get some effects:
+
+- effort to comprehend a piece of code will always grow more slowly than the complexity of that code, because:
+  - there's always going to be atleast one example of every potential usage, and
```
typo: atleast --> at<blank>least

jLouzado replied:
is it just me or do those two look the same? 🙈 

I think you mean `at-least` or `at least` instead.

dlebrero replied:
lol. I meant at least. I typed `at<blank>least` and Github decided to clean up my input.

dlebrero commented on content/blog/2019-10-3-tdd-gimmick/index.md:
```diff
@@ -0,0 +1,283 @@
+
+- effort to comprehend a piece of code will always grow more slowly than the complexity of that code, because:
+  - there's always going to be atleast one example of every potential usage, and
+  - you get an opportunity to play around with the code and see how it affects the test-suite.
```
test-suite --> `test suite`

dlebrero commented:
On every save is perfect, but this is something that people that do not
TDD is not aware of. I think it is important to make it clear for them
when/how often the tests should be run.

On 13/10/2019 09:43, Joel Louzado wrote:
>
> *@jLouzado* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In content/blog/2019-10-3-tdd-gimmick/index.md
> <https://github.com/jLouzado/patchcording.com/pull/11#discussion_r334266795>:
>
> > +So confidence (via code-coverage) is the whole game; you will see the above benefits in your codebase to the extent that the requirements are covered, simple as that.
> +
> +## What does "Produced By TDD" mean?
> +
> +First, a quick recap of TDD's rules.
> +
> +- If you want to write new code (or make changes to an existing unit) first write a failing test.
> +  - this is the `RED` phase.
> +- Only write as much code as is necessary to produce a failure.
> +  - compilation errors count as failures, and
> +- Only write as much code as is required to fix that failure, nothing more.
> +  - this is the `GREEN` phase
> +- Rinse and repeat till all your requirements are implemented.
> +- Go back over your implementation and clean-up.
> +  - This is the `REFACTOR` phase.
> +
>
> is there a prescribed time to run tests? I just set them to run on
> every save.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/jLouzado/patchcording.com/pull/11?email_source=notifications&email_token=AAQMJU4GMXS4ROST5LAUNDDQOLGQJA5CNFSM4I6QW3Q2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCHYXLMQ#discussion_r334266795>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAQMJUY4V74T7N4HTZLY7LDQOLGQJANCNFSM4I6QW3QQ>.
>

dlebrero commented:
"And why was that the requirement?" is the question that I now struggle
with. But the comment was just for me to raise a pain that I have seen,
which TDD does not solve. Maybe you have a good anwser to record the whys!

On 13/10/2019 09:53, Joel Louzado wrote:
>
> *@jLouzado* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In content/blog/2019-10-3-tdd-gimmick/index.md
> <https://github.com/jLouzado/patchcording.com/pull/11#discussion_r334267108>:
>
> > +
> +- it makes such an error easier to spot during code-review
> +  - since your `expected` value will clearly be something different than the requirement, someone can flag it as an issue
> +- once the error is spotted, it becomes easier to rewrite the bad test or to add tests for missing edge cases.
> +
> +Since refactoring is easier, this whole process happens much more quickly. And yes, while fixing one issue you're less likely to introduce a a regression even in this new feature.
> +
> +## This is all horribly complicated! I'm a pretty good programmer, I don't need all this
> +
> +I felt the same way when first starting out. Even now, when discussing some of the finer points of TDD it can feel a little overly-cautious and pessimistic. The thing is though, if there's more than 2 programmers working on a project it's just not possible to always be available for everyone.
> +
> +Imagine you've written a function, and someone wants to make a change to it. If they knew that the tests were high-quality, they could just figure it out themselves and be fairly confident that they weren't doing anything wrong. In the absence of a test-suite they would (hopefully) come and sit on your head and work things out with you.
> +
> +In the worst case, they're under a lot of pressure so they make the change that re-introduces some subtle bug that you'd already taken care of initially. The bug might only get discovered months later, at which point maybe the dev who introduced the problematic-change is on leave and the commit message just says "fixed issue" so now you're really screwed.
> +
> +At it's best, TDD is a way to communicate your decisions without needing to write extensive documentation that no one will ever read (or maintain when they make changes). Good-quality unit tests save such an incredible amount of time, and ultimately that's more time that you have to work in peace.
>
> Usually isn't the why just "because that was the requirement" ? So if
> the test description is written in terms of the product requirement
> then it should be easy to see what needs to be changed imo.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/jLouzado/patchcording.com/pull/11?email_source=notifications&email_token=AAQMJU3YVSA6FXJ2UKYEGZDQOLHWPA5CNFSM4I6QW3Q2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCHYXOAI#discussion_r334267108>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAQMJU2XMQT3Z6QXWKXQ6VLQOLHWPANCNFSM4I6QW3QQ>.
>

dlebrero commented:
I see what you mean now. Maybe I would change "you're almost in research
mode" for "you're in research mode"


On 13/10/2019 11:01, Joel Louzado wrote:
>
> *@jLouzado* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In content/blog/2019-10-3-tdd-gimmick/index.md
> <https://github.com/jLouzado/patchcording.com/pull/11#discussion_r334269202>:
>
> > +
> +In the worst case, they're under a lot of pressure so they make the change that re-introduces some subtle bug that you'd already taken care of initially. The bug might only get discovered months later, at which point maybe the dev who introduced the problematic-change is on leave and the commit message just says "fixed issue" so now you're really screwed.
> +
> +At it's best, TDD is a way to communicate your decisions without needing to write extensive documentation that no one will ever read (or maintain when they make changes). Good-quality unit tests save such an incredible amount of time, and ultimately that's more time that you have to work in peace.
> +
> +## Are there any cases where one shouldn't use TDD?
> +
> +For application development, almost everything is a good candidate for TDD. The one exception for frontend development would be if your function returns a particular DOM structure, there's no point writing a test like `Assert <div>Hello World!</div>` and then having a function that just returns `<div>Hello World!</div>`.
> +
> +You _could_ do it, but then tomorrow if you need to return `<div>Hello <b>World!</b></div>` should the test really break? I mean, I added an enhancement, the old test should "technically" still be valid.
> +
> +This doesn't mean that all view-related things cannot be done via TDD though. You might have a rock-solid reducer implemented, but if you forget to actually fire an action when a button is clicked then your app is still buggy. Packages like [react-testing-library](https://github.com/testing-library/react-testing-library) help with that problem.
> +
> +## Any others?
> +
> +Yes, if you're writing a new package, and there's two potential APIs that would both work equally fine but they both have different (say) performance implications based on their implementation then that might be a bad use-case for TDD.
>
> made a change, let me know if this is clearer.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/jLouzado/patchcording.com/pull/11?email_source=notifications&email_token=AAQMJU5BEFXTA3DQCTFFAATQOLPVPA5CNFSM4I6QW3Q2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCHYX7GA#discussion_r334269202>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAQMJU6ELDW4TBMVLDSZVGTQOLPVPANCNFSM4I6QW3QQ>.
>

jLouzado committed:
fix(tdd-gimmick): Fix `atleast` typo

fix: pull/11#discussion_r334362247

jLouzado committed:
refactor(tdd-gimmick): Clarify research mode and TDD

fix: pull/11#issuecomment-541665180

jLouzado committed:
fix(tdd-gimmick): Add note about communication via tests

fix: pull/11#discussion_r336726251

jLouzado committed:
fix(tdd): Cleanup titles

jLouzado committed:
fix(tdd): Remove use of word 'ideally'

People tend to hear Ideally and go "Yes, but in the real world..." as a
retort.

jLouzado committed:
fix(tdd): Improve ending

jLouzado committed:
refactor(tdd): Clarify point about buggy tests if writing tests second

jLouzado committed:
refactor(tdd): Move logic-errors section higher

jLouzado committed:
refactor(tdd): Move updating-old-tests down

jLouzado committed:
feat(tdd): Add section for designing the function-API

jLouzado committed:
fix(tdd): Add slight punch-up

jLouzado committed:
fix(tdd): Clarify what is meant by a flux app

jLouzado committed:
feat(tdd): Add description

jLouzado committed:
refactor(tdd): Move note on `why` of tests to footnote

jLouzado committed:
fix(tdd): Fix punctuation and punch-up language

jLouzado committed:
fix(tdd): Replace test-suite with 'test suite'

fix: pull/11#discussion_r334455978

jLouzado committed:
fix(tdd): Fix typo

fix: pull/11#commitcomment-35573657
